src/write_gcm.c

```diff
@@ -2147,6 +2169,52 @@ static monitored_resource_t *wg_monitored_resource_create(
   return NULL;
 }
 
+static monitored_resource_t *wg_monitored_resource_create_from_metadata_agent(
+    const char *metadata_agent_url, const char *project_id,
+    const char *resource_id) {
+  if (resource_id == NULL) {
+    WARNING("write_gcm: Hostname not set. Cannot query Metadata Agent.");
+    return NULL;
+  }
+  char *body = NULL;
+  const char **headers = NULL;
+  int num_headers = 0;
+  const char *url = metadata_agent_url != NULL ?
+      metadata_agent_url : metadata_agent_default_url;
+  char *response_buffer = (char *)
+      calloc(METADATA_RESPONSE_BUFFER_SIZE, sizeof(char));
+  if (response_buffer == NULL) {
+    ERROR("write_gcm: wg_monitored_resource_create_from_metadata_agent:"
+          "calloc failed.");
+    return NULL;
+  }
+  size_t url_size = strlen(url) + strlen(resource_id) +
+      strlen("/monitoredResource/") + 1;
+  {
+  char query[url_size];
+  int written = ssnprintf(query, url_size, "%s/monitoredResource/%s", url,
+                    resource_id);
+  DEBUG("write_gcm: wg_monitored_resource_create_from_metadata_agent: "
+        "Written: %d, Size: %zu, URL: %s", written, url_size, query);
+
+  // ssnprintf returns number of characters written excluding the \0 character
+  // but the size argument it takes requires \0 to be accounted for.
+  if (url_size - written != 1) {
+    ERROR("write_gcm: wg_monitored_resource_create_from_metadata_agent: %s",
+          "Could not create Metadata Agent URL.");
+    goto error;
+  }
+
+  wg_curl_get_or_post(response_buffer, METADATA_RESPONSE_BUFFER_SIZE, query,
+                      body, headers, num_headers);
```

Comment:
Those are called "explaining variables", BTW, and they're generally [encouraged](http://wiki.c2.com/?IntroduceExplainingVariable).